[RDY] vim-patch: 7.4.2212, 7.4.2237, 7.4.2309, 7.4.2312

gregorias opened the pull request:
Problem:    Crash when autocommand moves to another tab. (Dominique Pelle)
Solution:   When navigating to another window halfway the :edit command go
            back to the right window.

https://github.com/vim/vim/commit/5a49789a9b1f6447aeafbbbdd5b235dd10c471d5

gregorias committed:
vim-patch:7.4.2312

Problem:    Crash when autocommand moves to another tab. (Dominique Pelle)
Solution:   When navigating to another window halfway the :edit command go
            back to the right window.

https://github.com/vim/vim/commit/5a49789a9b1f6447aeafbbbdd5b235dd10c471d5

gregorias commented:
I'm not sure whether I should and how to port the additional tests introduced in this patch. Could you help me out?

KillTheMule commented:
The simplest thing would probably be to also port [patch 7.4.2227](https://github.com/vim/vim/commit/1381d791478ee77c8368b89a3d4954316bf839b5), which introduces `test_tabpage.vim`, so you can add your changes just like upstream. If you feel like doing a complete round, you could also port 2237 and 2309 with this batch (see the [history of `test_tabpage.vim`](https://github.com/vim/vim/commits/5a49789a9b1f6447aeafbbbdd5b235dd10c471d5/src/testdir/test_tabpage.vim)), but that's certainly not mandatory.

Shougo commented:
I think 7.4.2312, 2313, 2314 is set changes.

gregorias committed:
vim-patch:7.4.2212

Problem:    Mark " is not set when closing a window in another tab. (Guraga)
Solution:   Check all tabs for the window to be valid. (based on patch by
            Hirohito Higashi, closes vim/vim#974)

https://github.com/vim/vim/commit/e59215c7dcae17b03daf39517560cfaa03314f5a

gregorias commented:
According to the [wiki](https://github.com/neovim/neovim/wiki/Merging-patches-from-upstream-Vim) I shouldn't port this test file, so I'd like to skip this one.

I found that I actually need a different patch #5505, before merging this one. This PR doesn't even build right now, I should have checked it.

KillTheMule commented:
I'd say it'd be better both for reviewing as well as the git history if you included all those related patches in this PR here. This will also speed up the process somewhat. Feel free to force push to your feature branch freely :)

gregorias committed:
vim-patch:7.4.2237

Problem:    Can't use "." and "$" with ":tab".
Solution:   Support a range for ":tab". (Hirohito Higashi)

https://github.com/vim/vim/commit/9b7f8ce9eb3cb704f8cc14ab659bf86b1d6dc13c

gregorias committed:
vim-patch:7.4.2309

Problem:    Crash when doing tabnext in a BufUnload autocmd. (Dominique Pelle)
Solution:   When detecting that the tab page changed, don't just abort but
            delete the window where w_buffer is NULL.

https://github.com/vim/vim/commit/11fbc2866ccc11b4dd1726abdaf582a78ef3f743

gregorias committed:
Fix lint errors.

justinmk commented on src/nvim/testdir/test_tabpage.vim:
```diff
@@ -186,4 +186,50 @@ function Test_tabpage_with_autocmd()
+  bw!
+endfunction
+
+func Test_tabnext_on_buf_unload()
```
this was renamed to `Test_tabnext_on_buf_unload1()` in https://github.com/vim/vim/commit/5a49789a9b1f6447aeafbbbdd5b235dd10c471d5

and we are missing `Test_tabnext_on_buf_unload2()` from that patch.

justinmk replied:
I'll fix this while merging.

justinmk commented:
Verified (some tests). Merged.

@gregorias Thanks! Nice choices (such as FUNC_ attributes and shada lua test).
